src/Compilers/Core/Portable/CommandLine/AnalyzerConfig.cs

```diff
@@ -0,0 +1,485 @@
+                        {
+                            severity = ReportDiagnostic.Suppress;
+                        }
+                        else
+                        {
+                            severity = null;
+                            diagnosticBuilder.Add(Diagnostic.Create(
+                                InvalidAnalyzerConfigSeverityDescriptor,
+                                Location.None,
+                                diagId,
+                                value,
+                                analyzerConfigPath));
+                        }
+
+                        if (severity.HasValue)
+                        {
+                            treeBuilder[diagId] = severity.GetValueOrDefault();
+                        }
+                    }
+                    else
+                    {
+                        analyzerBuilder[key] = value;
+                    }
+                }
+            }
+        }
+
+        private static DiagnosticDescriptor InvalidAnalyzerConfigSeverityDescriptor
+            = new DiagnosticDescriptor(
+                "InvalidSeverityInAnalyzerConfig",
+                CodeAnalysisResources.WRN_InvalidSeverityInAnalyzerConfig_Title,
+                CodeAnalysisResources.WRN_InvalidSeverityInAnalyzerConfig,
+                "AnalyzerConfig",
+                DiagnosticSeverity.Warning,
+                isEnabledByDefault: true);
+
+        public Section GlobalSection { get; }
+
+        /// <summary>
+        /// The directory the editorconfig was contained in, with all directory separators
+        /// replaced with '/'.
+        /// </summary>
+        public string NormalizedDirectory { get; }
+
+        /// <summary>
+        /// The path passed to <see cref="AnalyzerConfig.Parse(string, string)"/> during construction.
+        /// </summary>
+        public string PathToFile { get; }
+
+        /// <summary>
+        /// Comparer for sorting <see cref="AnalyzerConfig"/> files by <see cref="NormalizedDirectory"/> path length.
+        /// </summary>
+        internal static Comparer<AnalyzerConfig> DirectoryLengthComparer { get; } = Comparer<AnalyzerConfig>.Create(
+            (e1, e2) => e1.NormalizedDirectory.Length.CompareTo(e2.NormalizedDirectory.Length));
+
+        public ImmutableArray<Section> NamedSections { get; }
+
+        private AnalyzerConfig(
+            Section globalSection,
+            ImmutableArray<Section> namedSections,
+            string pathToFile)
+        {
+            GlobalSection = globalSection;
+            NamedSections = namedSections;
+            PathToFile = pathToFile;
+
+            // Find the containing directory and normalize the path separators
+            string directory = Path.GetDirectoryName(pathToFile) ?? pathToFile;
+            NormalizedDirectory = PathUtilities.DirectorySeparatorChar == '/'
+                ? directory
+                : directory.Replace(PathUtilities.DirectorySeparatorChar, '/');
```

Comment:
This logic is repeated above in a local function. We should probably just add a helper method to `PathUtilities`